protocol/op-supervisor-integration.md

```diff
@@ -0,0 +1,307 @@
+- Where L2 blocks were derived from on L1: determining which L2 block needs to be reorged out upon invalid executing messages
+    requires an understanding of L1-ordering.
+- When we choose to go with a deposits-only block, it may invalidate other L2 data at the same timestamp.
+- Relations between L2 cross-safe increments, to set up a FP pre-state view.
+
+# Proposed Solution
+
+The solution consists of 3 parts:
+- super-safe-db: data in supervisor to resolve safety.
+- op-supervisor RPC: serve queries to op-node / op-geth.
+- integration: op-node / op-geth access op-supervisor RPC to maintain a view of safety.
+
+## Super-safe-db
+
+### Primary key
+
+With L1 block number as primary key:
+
+- Key-space may leave gaps in L2 blockhashes
+- Not superchain timestamp aligned
+- Binary search to find what L1 blocks we need to determine a specific L2 timestamp
+- Binary search to determine L2 safety
+- Easy to roll back on L1 reorg
+
+With superchain timestamp as primary key:
+
+- All L2 blockhashes can be found.
+- We store more duplicate derived-from information; presumably 1 block-number per L2 block, and the L1 blockhash of at least the highest L1 block number.
+- Binary-search to find what L2 data was derived from up to a specific L1 block.
+- More difficult to roll back on L1 reorg.
+- Easy to capture interop timestamp invariant, thus easy to maintain horizon properties.
+- Weakest-link safety problem: cross-safe becomes global, increments all at once for everyone.
+  (This forms a delayed safety risk when 1 chain is not batch-submitting).
+- Consistent, roll-back on invalidated executing messages always straight-forward.
+- While the same as the FP bisection step, the L1-view is now inconsistent,
+  or completely delayed/unavailable, between different L2 chains.
+
+With each L1-block or timestamp increment, we need to capture the addition
+of any L2 block that is failing executing-message checks. 
+
+Note that additions are dynamic-size (based on chain interop set), and so every lookup will always be some kind of search.
+
+### DB updates by op-supervisor
+
+This DB can then be used to query when each L2 block became cross-safe.
+
+This same DB can be used by the challenger, to replace the existing safe-db. A must when starting interop-FP work.
+
+And this same DB can be used to improve the `FindSyncStart` performance, since the super-safe-db will have the exact data to determine where to start syncing from, when given a L1 block and chainID.
+
+#### finalized (always cross)
+
+To query the latest finalized L2 block,
+simply binary search for the last L2 block fully derived from the finalized L1 block.
+This can simplify existing op-node functionality, and make finality more reliable.
+
+#### cross-safe
+
+To query whether an L2 block is `cross-safe`, binary search for the L2 block, to determine when (w.r.t. L1 block)
+it became cross-safe. If this L1 block exists and is canonical, then the L2 block is cross-safe.
```

Comment:
Just thinking through the two different requirements we have here:

1. We don't write an L2 block to the super-safe-db until it is cross safe and we write the L2 block that is cross safe for all chains in that L1 at the same time
2. We progress local-safe derivation on each chain L1 block by L1 block so the chains move forward and eventually all chains have local-safe blocks at the same timestamp so we can determine if all dependencies are met or if we need to replace with a deposit-only block.

Since chains may submit batches at very different intervals, it's possible that when supervisor is up to L1 block 10, chain A can derive the next block, but chain B can't yet. So supervisor has to move forward along L1, potentially up to a full sequencer window, trying to find the next block from chain B.  In the mean time, chain A might derive a lot of L2 blocks.

I think this might work if the entries for each L1 always describe the current cross-safe L2 block for the chain at that L1.  So chain A would record the same L2 block for all of those L1 blocks as it waits for the next chain B block.  supervisor would still index all those new blocks from chain A because it's indexing up to unsafe head.

I'm assuming that would be the case even if there were no dependencies from chain A blocks because we want to move the L2 blocks ahead in sync?

But I don't think this behaviour matches what's currently in this doc.  The supervisor would need to tell all chains to start processing the next L1 block and just track the latest block they've reported back for each block. It can record a new L1 block set of entries in the db once it has a response from each chain for that L1 block, even if they didn't derive a new head. So it has to track the L1 block its up to rather than always returning the L1 block of the current safe head, and it would return the L2 block the chain last reported it was up to.

Eventually as the L1 moves forward, op-supervisor winds up having a block from every chain that builds on top of that chain's current cross-safe. It can then check if dependencies are met and advance the safe head or start assigning tasks to replace local-safe blocks with deposit-only blocks. It would need to reset the L1 block its asking nodes to derive from back to when the cross-safe head last advanced. Except at that point the local-safe for chain A could still be ahead of cross-safe (ie the batches for those blocks are in L1 blocks _prior_ to when the cross-safe advanced even those they're after the cross-safe block for chain A). Those batches are probably invalid now but there may have been others that are now valid and were ignored because the local-safe head had already advanced. The fact that different channels can overlap makes it even more complex to know which L1 block we need to go back to.

Reply:
Just realised - with steady batch derivation we don't need to go back and look for possible other batches. If the first one turns out to be invalid it is replaced with a deposit-only block and any others for that same block remain invalid.  Will have to think it through again but that should let use rewind the L1 to the block the cross safe last updated and try and go forward from there again.

Reply:
> I think this might work if the entries for each L1 always describe the current cross-safe L2 block for the chain at that L1. So chain A would record the same L2 block for all of those L1 blocks as it waits for the next chain B block. supervisor would still index all those new blocks from chain A because it's indexing up to unsafe head.

> I'm assuming that would be the case even if there were no dependencies from chain A blocks because we want to move the L2 blocks ahead in sync?

Right, we don't want a "weakest link batch submission" problem, and that's why I pivoted from using a timestamp (as in original notion doc draft) to a L1 block as primary key.

If a L2 chain does not make progress on safety, it just repeats the latest cross-safe block in the DB, for those L1 blocks that it did not make progress for.
That way we have a notion, at any point in time in L1, where the L2s were at in terms of derived blocks.

Upon reorgs of L1, it'll be easy to roll back the DB by primary key, towards a L1 block that is still there, and get a coherent older understanding of where L2s were at. And then continue from there, to follow the new L1 chain.

Also note that we are storing both the "local" safe and "cross" safe L2-pointers per L1 block, this makes it easy to promote a "local" safe once the dependencies are checked.

> I'm assuming that would be the case even if there were no dependencies from chain A blocks because we want to move the L2 blocks ahead in sync?

Yes, no weakest-link, the DB should always progress. If the L1 did not produce new L2 blocks, then just repeat the last known L2 block (for cross-safe, and also local-safe, labels).

If derivation is not being processed, i.e. if we are unable to determine what L2 block is being derived from L1 at all, then we are forced to stop the DB progression however. This is unavoidable, because inactive op-nodes cannot be attributed to L1, so any action would be premature.

> But I don't think this behaviour matches what's currently in this doc. The supervisor would need to tell all chains to start processing the next L1 block and just track the latest block they've reported back for each block. It can record a new L1 block set of entries in the db once it has a response from each chain for that L1 block, even if they didn't derive a new head. So it has to track the L1 block its up to rather than always returning the L1 block of the current safe head, and it would return the L2 block the chain last reported it was up to.

The behavior you are describing here matches my understanding, and what I tried to express in the doc.

> Eventually as the L1 moves forward, op-supervisor winds up having a block from every chain that builds on top of that chain's current cross-safe. It can then check if dependencies are met and advance the safe head or start assigning tasks to replace local-safe blocks with deposit-only blocks. It would need to reset the L1 block its asking nodes to derive from back to when the cross-safe head last advanced. Except at that point the local-safe for chain A could still be ahead of cross-safe (ie the batches for those blocks are in L1 blocks prior to when the cross-safe advanced even those they're after the cross-safe block for chain A). Those batches are probably invalid now but there may have been others that are now valid and were ignored because the local-safe head had already advanced. The fact that different channels can overlap makes it even more complex to know which L1 block we need to go back to.

The steady-batch-derivation solves the problem you outlined here. Once we invalidate a "local safe" block that was not promoted to cross-safe yet, while staying on the same L1 chain, it means we do not have to reconsider prior batches anymore. Once invalid, always invalid, when anchored to the same L1 chain. "steady", no retries, no if-this-then-that backtracking in L1 history.

> Just realised - with steady batch derivation we don't need to go back and look for possible other batches. If the first one turns out to be invalid it is replaced with a deposit-only block and any others for that same block remain invalid. Will have to think it through again but that should let use rewind the L1 to the block the cross safe last updated and try and go forward from there again.

Exactly. We can always drop the "local" safe, revert to the last "cross safe", while progressing on the same L1 chain without re-processing older L1 blocks. This does mean the "local safe" of the L2 is reorged, but the event-logs-DB can be updated, by rolling it back to cross-safe, and syncing the new post-interop-reorg L2 blocks.